src/ResourceManager/Common/Commands.ResourceManager.Common/ResourceGroupCompleter.cs

```diff
@@ -0,0 +1,134 @@
+//
+// Copyright Microsoft Corporation
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ----------------------------------------------------------------------------------
+
+namespace Microsoft.Azure.Commands.ResourceManager.Common.TabCompletion
+{
+    using Commands.Common.Authentication;
+    using Commands.Common.Authentication.Abstractions;
+    using Management.Internal.Resources;
+    using Management.Internal.Resources.Models;
+    using Properties;
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Management.Automation;
+
+    /// <summary>
+    /// This attribute will allow the user to autocomplete the -ResourceGroup parameter of a cmdlet with valid resource groups
+    /// </summary>
+    public class ResourceGroupCompleterAttribute : ArgumentCompleterAttribute
+    {
+        private static IList<String> _resourceGroupNames = new List<string>();
+        private static readonly object _lock = new object();
+
+        protected static IList<String> ResourceGroupNames
+        {
+            get
+            {
+                lock (_lock)
+                {
+                    _resourceGroupNames = new List<string>();
+                    IAzureContext context = AzureRmProfileProvider.Instance.Profile.DefaultContext;
+                    try
+                    {
+                        var instance = AzureSession.Instance;
+                        var client = instance.ClientFactory.CreateCustomArmClient<ResourceManagementClient>(
+                            context.Environment.GetEndpointAsUri(AzureEnvironment.Endpoint.ResourceManager),
+                            instance.AuthenticationFactory.GetServiceClientCredentials(context, AzureEnvironment.Endpoint.ResourceManager),
+                            instance.ClientFactory.GetCustomHandlers());
+                        //var client = AzureSession.Instance.ClientFactory.CreateArmClient<ResourceManagementClient>(context, AzureEnvironment.Endpoint.ResourceManager);
+                        client.SubscriptionId = context.Subscription.Id;
+                        var resourceGroups = client.ResourceGroups.ListAsync();
+                        if (resourceGroups.Wait(TimeSpan.FromSeconds(5)))
+                        {
+                            if (resourceGroups.Result != null)
+                            {
+                                var resourceGroupList = resourceGroups.Result.ToList();
```

Comment:
Why write it to a list instead of just using it in the ForEach?

Reply:
Also, here we are making a conscious decision to only use the first page of results each time.  We should probably call this out with a comment.  BTW - ho9w many resource groups are in the first page of results?

Reply:
I couldn't figure out how many are in the first page - it wasn't in any of the code or documentation about IPage or List and it's more than my biggest subscription (92).  Not sure if there's another place I should look.